cjwstate/clientside.py

```diff
@@ -196,6 +197,10 @@ class StepUpdate:
     If the cached-result delta ID is different from the _relevant_ delta ID,
     then the result is obsolete. The client may render obsolete data plus a
     progress indicator.
+    
+    When a `render_result` is provided, you must also pass a `module_slug`.
+    This is so that `I18nMessage`s with source `"module"` in the result
+    can be interpreted relative to their source module.
```

Comment:
Please add a `__post_init__()` to assert this is the case. Otherwise we can easily regress. (As you've seen, the error is tricky to spot and to debug.)

`if self.render_result is not None: assert self.module_slug is not None`